docs: Guide for configuring automatic review rules

bernardjkim opened the pull request:
Supports: https://github.com/gravitational/teleport/issues/51682

This PR adds a guide to configuring automatic review rules via the Teleport Web UI.

bernardjkim committed:
Automatic reviews guide

bernardjkim committed:
lint

bernardjkim committed:
Merge branch 'master' into bernard/automatic-review-guide

bernardjkim committed:
Remove redundant sentence

ptgott commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
````diff
@@ -0,0 +1,134 @@
+$ tctl create demo-role.yaml
+```
+
+### Create the User
````
```suggestion
### Create the user
```
"user" is left lowercased in the docs

ptgott commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,134 @@
+
+  (!docs/pages/includes/add-role-to-user.mdx role="demo-access-request" user="\`demo-access-requester\`"!)
+
+## Step 2/4. Assign User Traits
```
```suggestion
## Step 2/4. Assign user traits
```

ptgott commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,134 @@
+   - **Review decision**: `APPROVED`
+4. Click **Create Rule**
+
+![Create Automatic Review Rule Editor](../../../../img/access-controls/automatic-reviews/create-automatic-review-rule.png)
```
The vertical orientation of this image makes it pretty large on the rendered docs page.

ptgott commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,134 @@
+role will be automatically reviewed for approval if the Teleport user traits
+requirements are satisfied. In this case, any user with the traits `team: demo`
+will satisfy the requirement.
+
```
Is the value of the **Key** field hardcoded, or does the user type it in? Does **Key** or **Value** support wildcards/regular expressions?

bernardjkim replied:
The key field is not hardcoded. The user can specify any value. Wildcards/regular expressions are not supported for key or value in this format.

ptgott commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,134 @@
+requirements are satisfied. In this case, any user with the traits `team: demo`
+will satisfy the requirement.
+
+## Step 4/4. Verify Automatic Review Rule
```
```suggestion
## Step 4/4. Verify the automatic review rule
```
Adding a missing article and using consistent capitalization (looks like "automatic review rule" is lowercased elsewhere).

ptgott commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,134 @@
+4. Add the `access` role to the Access Request
+5. Click **Proceed to Request**, then **Submit Request**
+
+![Create Access Request](../../../../img/access-controls/automatic-reviews/create-access-request.png)
```
This image is also pretty large on the rendered page.

ptgott reviewed (commented):
Minor comments

bernardjkim committed:
Merge branch 'master' into bernard/automatic-review-guide

bernardjkim committed:
Address feedback

bernardjkim committed:
Merge branch 'master' into bernard/automatic-review-guide

bernardjkim commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,134 @@
+
+(!docs/pages/includes/commercial-prereqs-tabs.mdx!)
+
+## Step 1/4. Create Requester Role and User
```
Oh, thanks for sharing this guide.

bernardjkim committed:
Resize images

bernardjkim committed:
Update terraform example

- Add desired_state spec.
- Replace automatic_approval with automatic_review spec.

bernardjkim committed:
Uppercase APPROVED and DENIED state

ptgott commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,134 @@
+4. Click **Save**
+5. Verify that the user has been updated with the desired trait.
+
+![Edit User Traits](../../../../img/access-controls/automatic-reviews/users-edit.png)
```
```suggestion
![Edit User Traits](../../../../img/access-controls/automatic-reviews/users-edit.png)

When adding user traits, you can enter any keys and values. The user trait form does not support wildcards or regular expressions.
```

Suggested because (as far as I recall) some Web UI forms do support these, so I thought we could make the constraints explicit.

bernardjkim committed:
Add IG prerequisite

bernardjkim committed:
Add access_monitoring_rules reference

bernardjkim committed:
Comment about wildcards and regex

bernardjkim committed:
make -C integrations/terraform docs

bernardjkim committed:
typo

zmb3 commented on docs/img/access-controls/automatic-reviews/create-access-request.png:
Is there an existing access request image in our docs that you can re-use here? I don't see anything specific to automatic reviewers here, so maybe we can avoid an extra screenshot.

zmb3 commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+description: Describes how to configure Access Monitoring Rules for Automatic Reviews.
+---
+
+Teleport supports automatic reviews of Access Requests for Teleport roles. This
```
> for Teleport roles.

Does this mean we don't support automatic reviews for resource access requests?

Edit: I now see this mentioned at the very end in the troubleshooting section. I feel like we should make this limitation more clear up front.

bernardjkim replied:
I've updated the first section to include this notice:
```
Support for automatic reviews of resource Access Requests is currently in development.
```

zmb3 commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+
+(!docs/pages/includes/commercial-prereqs-tabs.mdx!)
+
+- Teleport Identity Governance enabled on your account.
```
It's not clear that "account" here refers to your Teleport subscription and not your user account. What if we just said:
```suggestion
- This feature requires Teleport Identity Governance.
```

zmb3 commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+
+Automatic review rules can automatically approve or deny Access Requests based
+on the selected review decision. If an Access Request meets the conditions for
+both an approval rule and a denial rule, neither rule will be applied.
+
+If you suspect this issue, review your Access Monitoring Rules in the Teleport
+Web UI or by running `tctl get access_monitoring_rules`.
```
A few comments.

1. It's an interesting decision to apply neither. Typically in Teleport's RBAC system when multiple rules disagree then the strictest option wins. Any particular reason we made automatic reviews work differently?

2. From a UX perspective, "manually review all of your access monitoring rules to see if there's a conflict" is not very friendly. Can we detect this case and communicate the conflict to the user? (Not saying we need to do it now, but maybe we can create an issue to follow up).

bernardjkim replied:
> Typically in Teleport's RBAC system when multiple rules disagree then the strictest option wins.

That probably makes more sense. I'll create a follow up item make this change.

> From a UX perspective, "manually review all of your access monitoring rules to see if there's a conflict" is not very friendly.

Yeah, this is a concern that I've been thinking about how to address. We don't have a great solution for validating or reviewing access_monitoring_rules right now. The use of Teleport's predicate language makes this problem a bit complicated.

bernardjkim replied:
https://github.com/gravitational/teleport/pull/54933 changes the behavior when conflicting rules match a request. Denied reviews will now take precedence.

zmb3 commented on docs/pages/reference/resources.mdx:
```diff
@@ -568,3 +569,76 @@ status:
+kind: access_monitoring_rule
+version: v1
+metadata:
+  name: access_monitoring_rule
```
I think it would be helpful to come up with an example name which is not the same as the `kind` to help illustrate that these are separate values.

bernardjkim committed:
Update wording

bernardjkim committed:
reuse screenshot

bernardjkim committed:
Merge branch 'master' into bernard/automatic-review-guide

bernardjkim committed:
Edit conflicting rules behavior

bernardjkim committed:
Merge branch 'master' into bernard/automatic-review-guide

bernardjkim commented:
Hey @r0mant can you take a look when you have time?

We talked about adding a more in-depth reference with multiple different examples. I'll address this in a separate PR and structure it as a reference for `access_monitoring_rules`, with examples of both notification routing rules and automatic review rules.

r0mant commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+Teleport supports automatic reviews of role Access Requests. This
+feature enables teams to enforce a zero standing privilege policy, while
+still allowing users to receive temporary access without manual approval.
+
+Support for automatic reviews of resource Access Requests is currently in
+development.
```
I would just remove this.
```suggestion
```

bernardjkim replied:
I think it is worth mentioning up front that automatic reviews with resource requests are not supported yet. It was pointed out by Zac, and some other while testing this feature.

r0mant replied:
Let's make it an "info" admonition then at least? Otherwise it's just lost among the other text and people will skim over it.

r0mant commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+## How it works
+
+Automatic reviews are triggered by Access Monitoring Rules. These rules instruct
+Teleport to monitor specific Access Requests and automatically submit a review
```
```suggestion
Teleport to monitor Access Requests and automatically submit a review
```

r0mant commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+
+Automatic reviews are triggered by Access Monitoring Rules. These rules instruct
+Teleport to monitor specific Access Requests and automatically submit a review
+when certain conditions are met. The conditions can be configured based on the
+requested roles, and based on the requesting user's Teleport traits.
```
Nit: I would rephrase this a bit so we don't have to change this when we add more conditions.
```suggestion
when certain conditions (such as requested roles or user traits) are met.
```

r0mant commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+when certain conditions are met. The conditions can be configured based on the
+requested roles, and based on the requesting user's Teleport traits.
+
+For example, an Access Monitoring Rule can configure automatic Access Request
```
```suggestion
For example, an Access Monitoring Rule can perform an automatic Access Request
```

r0mant commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+requested roles, and based on the requesting user's Teleport traits.
+
+For example, an Access Monitoring Rule can configure automatic Access Request
+approval when a user with the Teleport traits `team: demo` requests access to
```
```suggestion
approval when a user with the Teleport trait or IdP attribute `team: demo` requests access to
```

r0mant commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
````diff
@@ -0,0 +1,144 @@
+$ tctl create demo-role.yaml
+```
+
+### Create the user
````
I'm wondering, should we split this section in 2 tabs that explains how to do this for a Teleport local vs SSO user?

- For Teleport local user, we tell them to create it and assign traits like you already have here.
- For SSO users, we tell them to ensure that IdP sends the required attribute (which will depend on IdP).

The reason is I think most people will want to use this feature with SSO users.

bernardjkim replied:
I think we should keep this guide simple and straight forward with minimal configuration options. Following the [styles guide](https://www.notion.so/goteleport/Style-guide-160fdd3830be80a79e70f08df2df2a0a?pvs=4#160fdd3830be8098bd44cfc5361a9c57), I think it'd be better to have a separate access_monitoring_rules reference page that provides examples for all these different types of configurations, including SSO users.

r0mant replied:
Sounds good. I think I would still at least mention here that if you're using SSO, then you can match on their attributes passed from an IdP. It's a useful context and most people reading this will be wondering.

r0mant commented on docs/pages/admin-guides/access-controls/access-requests/automatic-reviews.mdx:
```diff
@@ -0,0 +1,144 @@
+## Step 3/4. Create Access Monitoring Rule
+
+Next, define the automatic review rule via the Teleport Web UI.
+
```
Should we at least give a link here to Reference section here saying something like here's the resource spec if you want to create it via CLI? Or are you planning to add a section below that shows CLI/Terraform/etc. as a follow up?

bernardjkim replied:
I've included a link to the resource references page in the `Next Steps` section. I can also include more CLI/Terraform examples in a separate access_monitoring_rules reference page.

bernardjkim committed:
Minor language improvements

bernardjkim committed:
Add link to terraform resources index

r0mant reviewed (approved):
lgtm with a couple remaining suggestions

bernardjkim committed:
Address feedback

- Replace resource request notice with admonition
- Add SSO user admonition
